libminifi/include/utils/ResourceQueue.h

```diff
@@ -0,0 +1,132 @@
+ * The lent out resource is in a ResourceWrapper that returns the resource to the queue on destruction.
+ * */
+
+template<class ResourceType>
+class ResourceQueue : public std::enable_shared_from_this<ResourceQueue<ResourceType>> {
+ public:
+  class ResourceWrapper {
+   public:
+    ResourceWrapper(std::weak_ptr<ResourceQueue<ResourceType>> queue, std::unique_ptr<ResourceType>&& resource) : queue_(std::move(queue)), resource_(std::move(resource)) {}
+    ResourceWrapper(ResourceWrapper&& src) = default;
+    ResourceWrapper(const ResourceWrapper&) = delete;
+    ~ResourceWrapper() {
+      if (auto queue = queue_.lock())
+        queue->returnResource(std::move(resource_));
+    }
+
+    ResourceWrapper& operator=(ResourceWrapper&&) = default;
+    ResourceWrapper& operator=(const ResourceWrapper&) = delete;
+
+    ResourceType& operator*() const { return *resource_; }
+    ResourceType* operator->() const noexcept { return resource_.operator->(); }
+    ResourceType* get() const { return resource_.get(); }
+
+   private:
+    std::weak_ptr<ResourceQueue<ResourceType>> queue_;
+    std::unique_ptr<ResourceType> resource_;
+  };
+
+  static auto create(std::optional<size_t> maximum_number_of_creatable_resources, std::shared_ptr<core::logging::Logger> logger);
+
+  template<typename Fn>
+  [[nodiscard]] std::enable_if_t<std::is_invocable_v<std::unique_ptr<ResourceType>()>, ResourceWrapper> getResource(const Fn& create_resource) {
+    std::unique_ptr<ResourceType> resource;
+    // Use an existing resource, if one is available
+    if (internal_queue_.tryDequeue(resource)) {
+      logDebug("Using available [%p] resource instance", resource.get());
+      return ResourceWrapper(this->weak_from_this(), std::move(resource));
+    } else {
+      const std::lock_guard<std::mutex> lock(counter_mutex_);
+      if (!maximum_number_of_creatable_resources_ || resources_created_ < maximum_number_of_creatable_resources_) {
+        ++resources_created_;
+        resource = create_resource();
+        logDebug("Created new [%p] resource instance. Number of instances: %d%s.",
+                 resource.get(),
+                 resources_created_,
+                 maximum_number_of_creatable_resources_ ? " / " + std::to_string(*maximum_number_of_creatable_resources_) : "");
+        return ResourceWrapper(this->weak_from_this(), std::move(resource));
+      }
+    }
+    logDebug("Waiting for resource");
+    if (!internal_queue_.dequeueWait(resource)) {
+      throw std::runtime_error("No resource available");
+    }
+    return ResourceWrapper(this->weak_from_this(), std::move(resource));
+  }
+
+ protected:
+  ResourceQueue(std::optional<size_t> maximum_number_of_creatable_resources, std::shared_ptr<core::logging::Logger> logger)
+      : maximum_number_of_creatable_resources_(maximum_number_of_creatable_resources),
+        logger_(std::move(logger)) {
+  }
+
+ private:
+  void returnResource(std::unique_ptr<ResourceType>&& resource) {
```

Comment:
is this `&&` necessary?

Reply:
nope, I've removed it in https://github.com/apache/nifi-minifi-cpp/pull/1383/commits/7da01d53dfe1e127c2173fcbc9b9a9fadc712ff3